builder/builder.go

```diff
@@ -422,37 +432,44 @@ func (b *Builder) buildpackLayer(dest string, bp buildpack.Buildpack) (string, e
 
 	if err := tw.WriteHeader(&tar.Header{
 		Typeflag: tar.TypeDir,
-		Name:     buildpacksDir + "/" + bp.EscapedID(),
+		Name:     unixJoin(buildpacksDir, bp.EscapedID()),
 		Mode:     0755,
 		ModTime:  now,
 	}); err != nil {
 		return "", err
 	}
 
+	baseTarDir := unixJoin(buildpacksDir, bp.EscapedID(), bp.Version)
 	if err := tw.WriteHeader(&tar.Header{
 		Typeflag: tar.TypeDir,
-		Name:     buildpacksDir + "/" + bp.EscapedID() + "/" + bp.Version,
+		Name:     baseTarDir,
 		Mode:     0755,
 		ModTime:  now,
 	}); err != nil {
 		return "", err
 	}
 
-	if err := archive.WriteDirToTar(
-		tw,
-		bp.Dir,
-		fmt.Sprintf("%s/%s/%s", buildpacksDir, bp.EscapedID(), bp.Version),
-		b.UID,
-		b.GID,
-		false,
-	); err != nil {
+	if filepath.Ext(bp.Path) == ".tgz" {
```

Comment:
I wonder if we can have the buildpack object provide a TarReader in both the `.tgz` and dir case so we don't have to worry about dirs vs archives in so many separate places

Reply:
Deferred